File: heads/haskell/src/main/haskell/Hydra/Lib/Pairs.hs
-- | Haskell implementations of hydra.lib.pairs primitives

module Hydra.Lib.Pairs where

import qualified Data.Bifunctor as BF


-- | Map over both elements of a pair.
bimap :: (a -> c) -> (b -> d) -> (a, b) -> (c, d)
bimap = BF.bimap

-- | Get the first element of a pair.
first :: (a, b) -> a
first (x, _) = x

-- | Get the second element of a pair.
second :: (a, b) -> b
second (_, y) = y

File: heads/scala/src/main/scala/hydra/lib/pairs.scala
package hydra.lib

object pairs:
  def bimap[A, B, C, D](f: A => C)(g: B => D)(p: (A, B)): (C, D) = (f(p._1), g(p._2))
  def first[A, B](p: (A, B)): A = p._1
  def second[A, B](p: (A, B)): B = p._2
